### PR TITLE
Upgrade to lodash 4

### PR DESCRIPTION
**Pull request by islemaster**
Upgrades lodash from 3.10.0 to 4.13.1.

I carefully worked through the [4.0.0 upgrade guide](https://github.com/lodash/lodash/wiki/Changelog#v400), though by itself that wasn't sufficient to fix all of the apps tests; there were a few later changes that broke things too.
# Affected by the upgrade
- The optional `deep` argument to `_.clone()` has been removed; we now use `_.cloneDeep()` instead of `_.clone(_, true)`.
- The one-argument behavoir of `_.zipObject()` (accepting data in the format `[['key', 'value'], ['key', 'value']]`) has been moved to a new method `_.fromPairs`.
- `_.debounce` went through several changes; see below.
- The `_.contains` alias for `_.includes` has been removed.
- Certain lodash methods that accepted a callback used to follow the JavaScript convention of also accepting a context argument (e.g. the second arugment in `_.forEach(function () {...}, this)`).  This (already optional) context argument has been removed across the board - consumers are expected to bind their callbacks by some other means (like `Function.prototype.bind` or arrow functions).
# What's up with `_.debounce`?

A few things: The official breaking change in the 4.0 upgrade is that `_.debounce` no longer accepts a boolean as its third argument.  Way back [in Lo-Dash (sic) 1.0.0](https://github.com/lodash/lodash/blob/1.0.0/doc/README.md#_debouncefunc-wait-immediate) debounce's third argument was just a boolean `immediate` that invoked the wrapped function on the leading edge of the timeout instead of the trailing edge.  Equivalent behavior now passes an options object `{leading: true, trailing: false}`.  This was an easy fix.

However, `_.debounce` has been touched a ton since v4.0.0, sometimes in breaking ways:
- [4.3.0](https://github.com/lodash/lodash/wiki/Changelog#v430) says "Improved accuracy of `_.debounce` and `_.throttle`."
- [4.4.0](https://github.com/lodash/lodash/wiki/Changelog#v440) says "Ensured a debounced `maxWait` timeout isn’t processed on a leading call if `leading` is false & there isn’t a max delay queued". 
- [4.7.0](https://github.com/lodash/lodash/wiki/Changelog#v470) says "Refactored `_.debounce` to simplify, reduce timers, & fix bugs"
- Worst of all, [4.11.1](https://github.com/lodash/lodash/wiki/Changelog#v4111) says "Ensured `_.debounce` doesn’t immediately call func when wait is 0."  This a breaking change for us, it broke our tests where I often pass zero to bypass request coalescing and throttling in NetSim unit tests.  Before, calling `_.debounce(func, 0)()` would immediately execute func; now it behaves like `setTimeout(func, 0)` and doesn't invoke until the next tick.  I updated my tests to expect this (basically, doing `setTimeout(func, 0)` before verifying the effects of the debounced method) and we are back up and running.  In practice, outside of tests this change shouldn't have any effect.

### CODE REVIEW

**Comment by islemaster** [comment]
Note: I [opened an issue on lodash/lodash](https://github.com/lodash/lodash/issues/2524) discussing the "breaking" (for us anyway) change to `_.debounce` in 4.11.1.

File: apps/src/netsim/NetSimBitLogPanel.html.ejs [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <%
-  var i18n = require('./locale');
+  var i18n = require('@cdo/netsim/locale');
```
Comment (Bjvanminnen) [inline apps/src/netsim/NetSimBitLogPanel.html.ejs:2]: Is this related to lodash? Do we have guidance on when to use @cdo vs using a relative path?

File: apps/test/util/netsim/locale-do-not-import.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+ * DO NOT IMPORT THIS DIRECTLY. Instead do:
+ *   ```
+ *   import msg from '@cdo/netsim/locale'.
+ *   ```
+ * This allows the webpack config to determine how locales should be loaded,
+ * which is important for making locale setup work seamlessly in tests.
+ */
+// make sure Blockly is loaded
+require('../frame')();
+require('../../../build/package/js/en_us/netsim_locale.js');
```
Comment (Bjvanminnen) [inline apps/test/util/netsim/locale-do-not-import.js:11]: This means `npm run test` depends on `npm run build`. Was that the case before?

Comment (islemaster) [inline apps/test/util/netsim/locale-do-not-import.js:11]: It shouldn't; `newer:messages` (which I believe generates the locale files) is [a dependency of the `test` tasks](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/Gruntfile.js#L675), and AFAIK it has been for a while.

Comment (Bjvanminnen) [inline apps/test/util/netsim/locale-do-not-import.js:11]: Ahh, nevermind then

**Comment by Bjvanminnen** [comment]
lgtm

File: apps/src/netsim/NetSimBitLogPanel.html.ejs (hunk 1 above)
Comment (islemaster) [inline apps/src/netsim/NetSimBitLogPanel.html.ejs:2]: This is an approach to including locale files introduced by @pcardune in #9618 (and applied globally in #9621) which lets us stop worrying about require order for locale files in tests.  Where before we had to do this:

``` javascript
var utils = require('../../util/testUtils.js');
// We must invoke setupLocales before requiring ModuleUnderTest,
// because it depends on the appropriate localization functions already
// being present on the blockly global.
utils.setupLocales();
var ModuleUnderTest = require('@cdo/apps/ModuleUnderTest');
```

Now we can just do this, because in tests the locale import _within_ `ModuleUnderTest` is aliased in tests to a version that does preloading if necessary.

``` javascript
var ModuleUnderTest = require('@cdo/apps/ModuleUnderTest');
```

This was especially important as we move into ES6-land because Babel hoists `import` statements (but not `require` statements), foiling any attempt to do work before an import while using ES6 syntax.

This particular commit makes the same strategy work for NetSim files that don't depend on common locale, only netsim locale.  This wasn't causing failures on full test runs because the locale loading in other tests was setting up all locale files, but when using `karma:entry` to run NetSim tests by themselves (which I was doing while debugging the `_.debounce` issues) the locale wasn't being loaded.

Comment (pcardune) [inline apps/src/netsim/NetSimBitLogPanel.html.ejs:2]: I wonder how much value we get out of having separate imports for separate locales used in the different apps folders. Like maybe we should have something like:

``` javascript
import {netsim as i18n} from '@cdo/locale';
```

We could still keep the actual locale files separate, and just import/export them from the main `@cdo/locale` module.
